overlord/snapstate/handlers.go

```diff
@@ -2049,6 +2057,55 @@ func (m *SnapManager) maybeDiscardNamespacesOnSnapdDowngrade(st *state.State, sn
 	return nil
 }
 
+// maybeRemoveAppArmorProfilesOnSnapdDowngrade must be called when we are about
+// to activate a different snapd version. It checks whether we are a snapd with
+// a vendored AppArmor and whether we are performing a downgrade to an older
+// snapd version and, if so, discards all apparmor profiles; failure to do so
+// could cause the older snapd to be unable to load the existing apparmor
+// profiles generated by the vendored AppArmor parser.  This method assumes that
+// the State is locked.
+func (m *SnapManager) maybeRemoveAppArmorProfilesOnSnapdDowngrade(st *state.State, snapInfo *snap.Info) error {
+	if snapInfo.Type() != snap.TypeSnapd || snapInfo.Version == "" {
+		return nil
+	}
+
+	// ignore if not preseeded yet or if preseeding
+	var seeded bool
+	err := st.Get("seeded", &seeded)
+	if errors.Is(err, state.ErrNoState) || !seeded || snapdenv.Preseeding() {
+		return nil
+	}
+
+	if apparmor_sandbox.ProbedLevel() == apparmor_sandbox.Unsupported {
+		// no apparmor means no parser features
+		return nil
+	}
+
+	feats, err := apparmor_sandbox.ParserFeatures()
+	if err != nil {
+		return err
+	}
+
+	if !strutil.ListContains(feats, "snapd-internal") {
+		return nil
```

Comment:
this should have a comment

Reply:
added in 181afded50d73d81809eb8acc5fb26ead45ce49a